client/beefy/src/aux_schema.rs

```diff
@@ -0,0 +1,427 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Schema for BEEFY state persisted in the aux-db.
+
+use crate::worker::PersistedState;
+use beefy_primitives::{
+	crypto::AuthorityId, BeefyApi, ValidatorSet, BEEFY_ENGINE_ID, GENESIS_AUTHORITY_SET_ID,
+};
+use codec::{Decode, Encode};
+use futures::{stream::Fuse, StreamExt};
+use log::{debug, info, trace};
+use sc_client_api::{backend::AuxStore, Backend, FinalityNotifications};
+use sc_network_gossip::GossipEngine;
+use sp_api::{BlockId, HeaderT, ProvideRuntimeApi};
+use sp_blockchain::{
+	Backend as BlockBackend, Error as ClientError, HeaderBackend, Result as ClientResult,
+};
+use sp_runtime::traits::{Block as BlockT, Zero};
+
+const VERSION_KEY: &[u8] = b"beefy_auxschema_version";
+const WORKER_STATE: &[u8] = b"beefy_voter_state";
+
+const CURRENT_VERSION: u32 = 1;
+
+/// Write voter state.
+pub(crate) fn write_voter_state<Block: BlockT, B: AuxStore>(
+	backend: &B,
+	state: &PersistedState<Block>,
+) -> ClientResult<()> {
+	trace!(target: "beefy", "🥩 persisting {:?}", state);
+	backend.insert_aux(&[(WORKER_STATE, state.encode().as_slice())], &[])
+}
+
+fn load_decode<B: AuxStore, T: Decode>(backend: &B, key: &[u8]) -> ClientResult<Option<T>> {
+	match backend.get_aux(key)? {
+		None => Ok(None),
+		Some(t) => T::decode(&mut &t[..])
+			.map_err(|e| ClientError::Backend(format!("BEEFY DB is corrupted: {}", e)))
+			.map(Some),
+	}
+}
+
+/// Load or initialize persistent data from backend.
+pub(crate) async fn load_persistent<B, BE, R>(
+	backend: &BE,
+	runtime: &R,
+	gossip_engine: &mut GossipEngine<B>,
+	finality: &mut Fuse<FinalityNotifications<B>>,
+	min_block_delta: u32,
+) -> ClientResult<PersistedState<B>>
+where
+	B: BlockT,
+	BE: Backend<B>,
+	R: ProvideRuntimeApi<B>,
+	R::Api: BeefyApi<B>,
+{
+	let version: Option<u32> = load_decode(backend, VERSION_KEY)?;
+
+	match version {
+		None => (),
+		Some(1) =>
+			if let Some(mut state) = load_decode::<_, PersistedState<B>>(backend, WORKER_STATE)? {
+				// Overwrite persisted data with newly provided `min_block_delta`.
+				state.set_min_block_delta(min_block_delta);
+				info!(target: "beefy", "🥩 Loading BEEFY voter state from db: {:?}.", state);
+				return Ok(state)
+			},
+		other =>
+			return Err(ClientError::Backend(format!("Unsupported BEEFY DB version: {:?}", other))),
+	}
+
+	// If no persisted state present, start at the latest session boundary with BEEFY finality.
+	// TODO: we should actually start at pallet-beefy "genesis", but we don't know when that was,
+	// so we walk back the chain up to a session length looking for a starting point.
+	migrate_from_version0(backend, runtime, gossip_engine, finality, min_block_delta).await
```

Comment:
This is the only part that's a bit weird to me. I think it would have been better if this function returned `Option<PersistedState>` and this initialization was performed elsewhere explicitly.